src/wallet/wallet.cpp

```diff
@@ -2429,165 +2429,83 @@ const CTxOut& CWallet::FindNonChangeParentOutput(const CTransaction& tx, int out
 
-        const CWalletTx *pcoin = output.tx;
+        // Filter by the min conf specs and add to vValue and calculate effective value
+        for (const COutput &output : vCoins)
+        {
+            if (!output.fSpendable)
+                continue;
 
-        if (output.nDepth < (pcoin->IsFromMe(ISMINE_ALL) ? nConfMine : nConfTheirs))
-            continue;
+            const CWalletTx *pcoin = output.tx;
 
-        if (!mempool.TransactionWithinChainLimit(pcoin->GetHash(), nMaxAncestors))
-            continue;
-
-        int i = output.i;
+            if (output.nDepth < (pcoin->IsFromMe(ISMINE_ALL) ? nConfMine : nConfTheirs))
+                continue;
 
-        CInputCoin coin = CInputCoin(pcoin, i);
+            if (!mempool.TransactionWithinChainLimit(pcoin->GetHash(), nMaxAncestors))
+                continue;
 
-        if (coin.txout.nValue == nTargetValue)
-        {
-            setCoinsRet.insert(coin);
-            nValueRet += coin.txout.nValue;
-            return true;
-        }
-        else if (coin.txout.nValue < nTargetValue + MIN_CHANGE)
-        {
-            vValue.push_back(coin);
-            nTotalLower += coin.txout.nValue;
-        }
-        else if (!coinLowestLarger || coin.txout.nValue < coinLowestLarger->txout.nValue)
-        {
-            coinLowestLarger = coin;
+            int i = output.i;
+            CInputCoin coin(pcoin, i);
+            coin.effective_value = coin.txout.nValue - (output.nInputBytes < 0 ? 0 : effective_fee.GetFee(output.nInputBytes));
+            // Only include outputs that are not negative effective value (i.e. not dust)
+            if (coin.txout.nValue > 0) {
+                coin.fee = output.nInputBytes < 0 ? 0 : effective_fee.GetFee(output.nInputBytes);
+                coin.long_term_fee = output.nInputBytes < 0 ? 0 : long_term_feerate.GetFee(output.nInputBytes);
+                vValue.push_back(coin);
+            }
         }
-    }
-
-    if (nTotalLower == nTargetValue)
-    {
-        for (const auto& input : vValue)
+        return SelectCoinsBnB(vValue, nTargetValue, cost_of_change, setCoinsRet, nValueRet, not_input_fees);
+    } else {
+        // Filter by the min conf specs and add to vValue
+        for (const COutput &output : vCoins)
         {
-            setCoinsRet.insert(input);
-            nValueRet += input.txout.nValue;
-        }
-        return true;
-    }
-
-    if (nTotalLower < nTargetValue)
-    {
-        if (!coinLowestLarger)
-            return false;
-        setCoinsRet.insert(coinLowestLarger.get());
-        nValueRet += coinLowestLarger->txout.nValue;
-        return true;
-    }
+            if (!output.fSpendable)
+                continue;
 
-    // Solve subset sum by stochastic approximation
-    std::sort(vValue.begin(), vValue.end(), CompareValueOnly());
-    std::reverse(vValue.begin(), vValue.end());
-    std::vector<char> vfBest;
-    CAmount nBest;
+            const CWalletTx *pcoin = output.tx;
```

Comment:
In commit "Have SelectCoinsMinConf and SelectCoins use BnB or Knapsack" (44657d0374d7fb25cca5cd098b27a5163f139de6)

I'm not sure having `pcoin` and `i` variables actually makes the code easier to read. It would seem simpler to drop these extra variables and just write `output.tx` and `output.i` directly.

Reply:
Done